src/main/java/hudson/remoting/jnlp/Main.java

```diff
@@ -159,6 +172,68 @@ public Engine createEngine() {
         if(jarCache!=null)
             engine.setJarCache(new FileSystemJarCache(jarCache,true));
         engine.setNoReconnect(noReconnect);
+        if (candidateCertificates != null && !candidateCertificates.isEmpty()) {
+            CertificateFactory factory;
+            try {
+                factory = CertificateFactory.getInstance("X.509");
+            } catch (CertificateException e) {
+                throw new IllegalStateException("Java platform specification mandates support for X.509", e);
+            }
+            List<X509Certificate> certificates = new ArrayList<X509Certificate>(candidateCertificates.size());
+            for (String certOrAtFilename : candidateCertificates) {
+                certOrAtFilename = certOrAtFilename.trim();
+                byte[] cert;
+                if (certOrAtFilename.startsWith("@")) {
```

Comment:
In the case of files it would be great to require 700 permissions on Unix. Just to ensure that certificate is not leakable